plugins/module_utils/helpers.py

```diff
@@ -129,8 +129,13 @@ def helper_compare_lists(l1, l2, diff_dict):
         return diff_dict
     for i, item in enumerate(l1):
         if isinstance(item, dict):
-            diff_dict.insert(i, {})
-            diff_dict[i] = helper_compare_dictionaries(item, l2[i], diff_dict[i])
+            for i2, item2 in enumerate(l2):
+                diff_dict2 = {}
+                diff_dict2 = helper_compare_dictionaries(item, item2, diff_dict2)
+                if len(diff_dict2) == 0:
+                    break
+            if len(diff_dict2) != 0:
+                diff_dict.insert(i, item)
```

Comment:
I am having trouble understanding this correctly. Maybe my brain is not working optimally this morning :rofl: 

So previously, if we found dict as an element of list, we compared it with the item in the list2 at the same position, guessing that it would be a dict as well.

The new logic will, for each item that is a dict in list1, iterate whole dict2 and compare the item1 (dict) with everything in list2 no matter what that object would be (number/string for example). If it finds a match `len(diff_dict2) == 0` then it doesn't continue anymore. Wouldn't this require something like this?

```suggestion
            for i2, item2 in enumerate(l2):
                diff_dict2 = {}
                if isinstance(item2, dict):
                    diff_dict2 = helper_compare_dictionaries(item, item2, diff_dict2)
                    if len(diff_dict2) == 0:
                        break
            if len(diff_dict2) != 0:
                diff_dict.insert(i, item)
```

Also the enumeration index `i2` seems to be unused at all and can be omitted.

Reply:
Oh I spent couple of night with this problem. Imagine you work on a module locally without pushing anything to GitHub  and running integration test on this module over and over again, then at some point tests that were passing start to fail. Obviously I did not destroy Zabbix containers to speed up testing just manually deleting the resouce in Zabbix UI before re-running the test. I worked on zabbix_action.py and the test `- name: test - create new action with multiple conditions (reorder)` suddenly failed. I discovered that it was not returning re-ordered list anymore thus assertion `that: zbxaction_conditions_reorder.changed is sameas True` started to fail. Re-create containers and everything starts working again. I did not know what to do and was not sure how Python treats an order in lists/dictionaries and went to packet capture and guess what - after approximately 12-13 runs the order of elements in the list returned by Zabbix changes! I guess it is some kind of DB optimization optimization (plan? I am not a DB expert). I looked more closely at what we are trying to do and what we are expecting and it turns out that indeed the order of the elements in the list does not mater (at least I have not found any such case in this collection).
So the code is doing:
```
for every element in list l1
  iterate through all elements in list l2
    if any element in list l2 equals to element in list l1 then break this iteration and move to the next element in list l1
```
I am happy to add your `if isinstance(item2, dict):` but it might lead to unexpected results - if we compared a dict from l1 to let's say str in l2 (which should never happen!) we would not produce any diff and I'd rather see Python's exeption than silently ignoring this difference. Let me know.

And yes, I will definitely remote index i2.